File: internal/domain/router/router.go
// Package router отвечает за распространение событий по топологическому графу.
package router

import (
	"github.com/TheCipherKeeper/cybercity-engine/internal/domain/models"
)

// PropagationRule — чистая функция, которая по событию, исходному узлу и ребру
// решает, нужно ли породить дочернее событие для соседа.
type PropagationRule func(event models.EventNode, source *models.TopologyNode, edge models.TopologyEdge, topology *models.TopologyGraph) *models.EventNode

// EventRouter — graph-aware propagation engine.
type EventRouter struct {
	rules []PropagationRule
}

// NewEventRouter создаёт роутер с дефолтными правилами.
func NewEventRouter(rules ...PropagationRule) *EventRouter {
	if len(rules) == 0 {
		rules = defaultRules()
	}
	return &EventRouter{rules: rules}
}

// Propagate возвращает дочерние события для всех соседей, затронутых
// исходным событием.
func (r *EventRouter) Propagate(event models.EventNode, sourceNode *models.TopologyNode, topology *models.TopologyGraph) []models.EventNode {
	var children []models.EventNode
	for _, edge := range topology.Neighbors(sourceNode.ID) {
		for _, rule := range r.rules {
			child := rule(event, sourceNode, edge, topology)
			if child != nil {
				children = append(children, *child)
			}
		}
	}
	return children
}

func defaultRules() []PropagationRule {
	return []PropagationRule{
		scanAlertRule,
		compromisePropagationRule,
		stateChangePropagationRule,
	}
}

func scanAlertRule(event models.EventNode, source *models.TopologyNode, edge models.TopologyEdge, topology *models.TopologyGraph) *models.EventNode {
	if event.EventType != models.EventTypeScan {
		return nil
	}
	if edge.Kind != "log-sink" && edge.Kind != "trusts" {
		return nil
	}
	noiseRaw, ok := event.Payload["noise_level"]
	if !ok {
		return nil
	}
	noise := toFloat(noiseRaw)
	if noise < 0.5 {
		return nil
	}

	child := event.SpawnChild(
		models.EventTypePropagation,
		models.SourceTypeEngine,
		source.ID,
		edge.Target,
		map[string]any{
			"kind":              "scan_alert",
			"original_event_id": event.EventID,
			"noise_level":       noise,
			"via":               edge.Kind,
		},
		event.Tick,
	)
	return &child
}

func compromisePropagationRule(event models.EventNode, source *models.TopologyNode, edge models.TopologyEdge, topology *models.TopologyGraph) *models.EventNode {
	if event.EventType != models.EventTypeCompromise {
		return nil
	}
	if edge.Kind != "trusts" && edge.Kind != "auth" && edge.Kind != "vendor-vpn" {
		return nil
	}
	severity := toFloat(event.Payload["severity"])
	if severity < 0.3 {
		severity = 0.3
	}
	if severity < 0.3 {
		return nil
	}

	child := event.SpawnChild(
		models.EventTypeAttack,
		models.SourceTypeEngine,
		source.ID,
		edge.Target,
		map[string]any{
			"kind":              "lateral_movement",
			"original_event_id": event.EventID,
			"severity":          severity * 0.8,
			"via":               edge.Kind,
		},
		event.Tick,
	)
	return &child
}

func stateChangePropagationRule(event models.EventNode, source *models.TopologyNode, edge models.TopologyEdge, topology *models.TopologyGraph) *models.EventNode {
	if event.EventType != models.EventTypeStateChange {
		return nil
	}
	if event.Payload["field"] != "status" {
		return nil
	}
	newStatus, ok := event.Payload["new_value"].(string)
	if !ok {
		return nil
	}
	if newStatus != "down" && newStatus != "compromised" {
		return nil
	}
	if edge.Kind != "api-call" && edge.Kind != "db-read" && edge.Kind != "db-write" && edge.Kind != "backup-of" && edge.Kind != "log-sink" {
		return nil
	}

	child := event.SpawnChild(
		models.EventTypeBackgroundEffect,
		models.SourceTypeEngine,
		source.ID,
		edge.Target,
		map[string]any{
			"kind":              "dependency_impact",
			"original_event_id": event.EventID,
			"status":            newStatus,
			"via":               edge.Kind,
		},
		event.Tick,
	)
	return &child
}

func toFloat(v any) float64 {
	switch n := v.(type) {
	case float64:
		return n
	case float32:
		return float64(n)
	case int:
		return float64(n)
	case int64:
		return float64(n)
	}
	return 0
}

File: src/cybercity_engine/router.py
"""Event router — propagates events through the topology graph.

Given an event and a topology node, the router decides which neighbouring
services should receive derived events and what those events look like.
"""

from __future__ import annotations

from collections.abc import Callable

from .models import EventNode, EventType, TopologyEdge, TopologyGraph, TopologyNode

__all__ = ["EventRouter"]


PropagationRule = Callable[[EventNode, TopologyNode, TopologyEdge, TopologyGraph], EventNode | None]


class EventRouter:
    """Graph-aware event propagation engine.

    Rules are pure functions that inspect an event, the source node and an
    outgoing edge, then optionally return a child event for the neighbour.
    """

    def __init__(self, rules: list[PropagationRule] | None = None) -> None:
        self.rules = rules or self._default_rules()

    def propagate(
        self,
        event: EventNode,
        source_node: TopologyNode,
        topology: TopologyGraph,
    ) -> list[EventNode]:
        """Return child events for all affected neighbours."""
        children: list[EventNode] = []
        for edge in topology.neighbors(source_node.id):
            target = topology.services.get(edge.target)
            if target is None:
                continue
            for rule in self.rules:
                child = rule(event, source_node, edge, topology)
                if child is not None:
                    children.append(child)
        return children

    @staticmethod
    def _default_rules() -> list[PropagationRule]:
        return [
            _scan_alert_rule,
            _compromise_propagation_rule,
            _state_change_propagation_rule,
        ]


def _scan_alert_rule(
    event: EventNode, source: TopologyNode, edge: TopologyEdge, topology: TopologyGraph
) -> EventNode | None:
    """If a noisy scan happens on a service, alert its log-sink / IDS neighbours."""
    if event.event_type != EventType.SCAN:
        return None
    if edge.kind not in {"log-sink", "trusts"}:
        return None
    noisy = event.payload.get("noise_level", 0.0)
    if noisy < 0.5:
        return None
    return event.spawn_child(
        event_type=EventType.PROPAGATION,
        source_type="engine",
        source_id=source.id,
        target_id=edge.target,
        payload={
            "kind": "scan_alert",
            "original_event_id": event.event_id,
            "noise_level": noisy,
            "via": edge.kind,
        },
    )


def _compromise_propagation_rule(
    event: EventNode, source: TopologyNode, edge: TopologyEdge, topology: TopologyGraph
) -> EventNode | None:
    """A compromised service may propagate to trusted neighbours."""
    if event.event_type != EventType.COMPROMISE:
        return None
    if edge.kind not in {"trusts", "auth", "vendor-vpn"}:
        return None
    severity = event.payload.get("severity", 0.5)
    if severity < 0.3:
        return None
    return event.spawn_child(
        event_type=EventType.ATTACK,
        source_type="engine",
        source_id=source.id,
        target_id=edge.target,
        payload={
            "kind": "lateral_movement",
            "original_event_id": event.event_id,
            "severity": severity * 0.8,
            "via": edge.kind,
        },
    )


def _state_change_propagation_rule(
    event: EventNode, source: TopologyNode, edge: TopologyEdge, topology: TopologyGraph
) -> EventNode | None:
    """When a service goes down, send a resource-impact event to dependents."""
    if event.event_type != EventType.STATE_CHANGE:
        return None
    if event.payload.get("field") != "status":
        return None
    new_status = event.payload.get("new_value")
    if new_status not in {"down", "compromised"}:
        return None
    if edge.kind not in {"api-call", "db-read", "db-write", "backup-of", "log-sink"}:
        return None
    return event.spawn_child(
        event_type=EventType.BACKGROUND_EFFECT,
        source_type="engine",
        source_id=source.id,
        target_id=edge.target,
        payload={
            "kind": "dependency_impact",
            "original_event_id": event.event_id,
            "status": new_status,
            "via": edge.kind,
        },
    )
